README Rewrite

ShayneP opened the pull request:
The README has been a bit of a mess for a while, so I rewrote it.

Highlights :
- Added a TOC
- Moved the breaking changes into their own section that is at the end
- Added a section on Bulk
- Split up "Getting Started" into more logical pieces, and have separate paths for GraphQL vs REST
- Highlighted the importance of GraphQL for clients

tiffanytse commented on README.md:
```diff
@@ -5,134 +5,61 @@ Shopify API
+    + [Ruby version](#ruby-version)
+  * [Installation](#installation)
+  * [Getting Started](#getting-started)
+    + [1) Create an app](#1--create-an-app)
```
All of these bookmarks don't link properly, there's an extra `-` after the number, should be `#1-create-an-app`

ShayneP replied:
Good catch, forgot about that. Thanks!

tiffanytse commented on README.md:
```diff
@@ -5,134 +5,61 @@ Shopify API
+  * [Thread safety](#thread-safety)
+  * [Bulk Operations](#bulk-operations)
+    + [Example](#example)
+      - [1) Start the bulk operation](#1--start-the-bulk-operation)
```
Same here

tiffanytse commented on README.md:
```diff
@@ -146,33 +73,63 @@ Or install via [gem](http://rubygems.org/)
+For a private app, you'll need the API_KEY and the PASSWORD; otherwise, you'll need the API_KEY and SHARED_SECRET.
 
-2. For a private App you just need to set the base site url as follows:
+   If you're not sure how to create a new application in the partner/store admin or if you're not sure how to generate the required credentials, you can [read the related Shopify docs](https://docs.shopify.com/api/guides/api-credentials) on the same.
```
Update url to shopify.dev

tiffanytse replied:
https://shopify.dev/tutorials/authenticate-a-private-app-with-shopify-admin

ShayneP replied:
This isn't the same guide, I'll have to go back and rewrite the section to account for the change.

ShayneP replied:
Done!

tiffanytse commented on README.md:
````diff
@@ -146,33 +73,63 @@ Or install via [gem](http://rubygems.org/)
    ```
 
-   That's it, you're done, skip to step 6 and start using the API!
+   That's it; you're done! Next, skip to step 6 and start using the API!
````
Don't think this needs to change to a semicolon here?

ShayneP replied:
They're both independent clauses, so a comma creates a comma splice. We could put a period instead, but I figure the flow is a bit better with the semicolon.

tiffanytse commented on README.md:
```diff
@@ -182,59 +139,41 @@ ShopifyAPI uses ActiveResource to communicate with the REST web service. ActiveR
-   * Ensure the provided hostname parameter is a valid hostname, ends with myshopify.com, and does not contain characters other than letters (a-z), numbers (0-9), dots, and hyphens.
-
-   If all security checks pass, the authorization code can be exchanged once for a permanent access token. The exchange is made with a request to the shop.
+   This method will save the token to the session object and return it. All fields returned by Shopify, other than the access token itself, are stored in the session's `extra` attribute. For a list of all fields returned by Shopify, read [our OAuth documentation](https://help.shopify.com/api/guides/authentication/oauth#confirming-installation). 
```
url update here too (https://shopify.dev/tutorials/authenticate-with-oauth#confirming-installation)

tiffanytse commented on README.md:
````diff
@@ -146,33 +73,63 @@ Or install via [gem](http://rubygems.org/)
    ```ruby
    shop_url = "https://#{API_KEY}:#{PASSWORD}@#{SHOP_NAME}.myshopify.com"
    ShopifyAPI::Base.site = shop_url
    ShopifyAPI::Base.api_version = '<version_name>' # find the latest stable api_version [here](https://help.shopify.com/api/versioning)
````
url update

tiffanytse commented on README.md:
```diff
@@ -182,59 +139,41 @@ ShopifyAPI uses ActiveResource to communicate with the REST web service. ActiveR
 
    * ``client_id`` – Required – The API key for your app
    * ``scope`` – Required – The list of required scopes (explained here: https://help.shopify.com/api/guides/authentication/oauth#scopes)
```
url update

tiffanytse replied:
https://shopify.dev/tutorials/authenticate-with-oauth#scopes

tiffanytse commented on README.md:
```diff
@@ -247,24 +186,58 @@ ShopifyAPI uses ActiveResource to communicate with the REST web service. ActiveR
+
+###### Note: the GraphQL client has improved and changed in version 9.0. See the [client documentation](docs/graphql.md) for full usage details and a [migration guide](docs/graphql.md#migration-guide).
+
+This library also supports Shopify's [GraphQL Admin API](https://help.shopify.com/api/graphql-admin-api)
```
https://shopify.dev/docs/admin-api/graphql/reference

tiffanytse commented on README.md:
```diff
@@ -344,60 +317,148 @@ gem install shopify_api_console
 
-Note: the GraphQL client has improved and changed in version 9.0. See the [client documentation](docs/graphql.md)
-for full usage details and a [migration guide](docs/graphql.md#migration-guide).
+ActiveResource is threadsafe as of version 4.1 (which works with Rails 4.x and above).
```
should this be marked up as code? `4.x`

tiffanytse commented on README.md:
```diff
@@ -344,60 +317,148 @@ gem install shopify_api_console
-  page += 1
-  products = ShopifyAPI::Product.find(:all, params: { limit: 50, page: page })
-  process_products(products)
+# Efficient: reads the file a single line at a time
```
Nice!

tiffanytse commented on README.md:
```diff
@@ -344,60 +317,148 @@ gem install shopify_api_console
-Page based pagination will be deprecated in the `2019-10` API version, in favor of the second method of pagination:
+## Pagination
+
+Shopify uses [Relative cursor-based pagination](https://help.shopify.com/en/api/guides/paginated-rest-results) to provide more than a single page of results. 
```
https://shopify.dev/tutorials/make-paginated-requests-to-rest-admin-api

tiffanytse commented on README.md:
```diff
@@ -437,11 +620,11 @@ or you can even use our automated rake task for docker:
-## Additional Resources
+# Additional Resources
 
 * [API Reference](https://help.shopify.com/api/reference)
```
https://shopify.dev/docs/admin-api/rest/reference

ShayneP commented:
Added links to the new docs so that they don't have to redirect, rewrote the authentication section to account for the new tutorial format, changed forum links to the new forums, and added GraphQL Reference link at the end!
